Clean up local unassigned shards at raptor node startup.

omegaga opened the pull request:
Addresses issue #7585 

Do local shards cleanup without waiting the interval time at node startup.

omegaga commented:
@AlekseiS

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -352,12 +356,19 @@ synchronized void cleanLocalShards()
-                .map(Map.Entry::getKey)
-                .collect(toSet());
+        Set<UUID> deletions;
+        if (startup) {
```
I'd rather create a new method to do unconditional cleanup. This method does several actions which are unnecessary at startup. By removing them you simplify the logic.

omegaga commented:
Addressed comments of @AlekseiS and @electrum 

Now at startup time we fetch local shards first, wait jitter time and then read assigned shard metadata.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -275,9 +267,35 @@ private synchronized void runBackupCleanup()
         }
     }
 
+    private synchronized void waitJitterTime()
```
I don't think this method needs to be synchronized

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -275,9 +267,35 @@ private synchronized void runBackupCleanup()
+    private synchronized void runLocalCleanupAtStartup()
+    {
+        try {
+            Set<UUID> local = getLocalShards();
```
I'd get the list outside of scheduled method in a blocking way. That way it's guaranteed to have only the files you care about. That's what David originally suggested.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -275,9 +267,35 @@ private synchronized void runBackupCleanup()
     private synchronized void runLocalCleanup()
     {
         try {
+            waitJitterTime();
```
this method is exported to be called from jmx. Your change will cause it to sleep before actual invocation.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -275,9 +267,35 @@ private synchronized void runBackupCleanup()
+        }
+    }
+
+    private synchronized void runLocalCleanupAtStartup()
```
think about exporting this method for jmx, but without wait time. Something like "force clean".

AlekseiS commented on presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestShardCleaner.java:
```diff
@@ -189,6 +189,64 @@ public void testDeleteOldShards()
+        Duration cleanTime = new ShardCleanerConfig().getLocalCleanTime();
+        ticker.increment(cleanTime.toMillis() - 1, MILLISECONDS);
+
+        // try non-startup mode. This should not clean any shards
```
I'm not sure this part is relevant to your test.

omegaga replied:
My original plan was to make sure `cleanLocalShardsAtStartup` and `cleanLocalShards` behave differently. When ticker elapses within the clean time threshold, `cleanLocalShards` should not delete the shard but `cleanLocalShardsAtStartup` does not have this condition. This behavior is the reason why we want a method other than `cleanLocalShards` so I add it in the test.

AlekseiS replied:
I see you point. However, the goal of this test is to verify expected behavior of the method. You don't need to prove the method's usefulness or compare it with other ones in this test.
I think you can create a setup with incremented ticker, but I don't think you should run another cleaning method inside this test.

omegaga commented:
@AlekseiS would you take a look at this?

AlekseiS commented:
I'll have a look today.

AlekseiS commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -251,17 +252,15 @@ private void scheduleBackupCleanup()
+            waitJitterTime();
             runLocalCleanup();
-        }, 0, localCleanerInterval.toMillis(), MILLISECONDS);
+        }, localCleanerInterval.toMillis(), localCleanerInterval.toMillis(), MILLISECONDS);
```
I think you don't need to delay the start of usual scheduler, because it significantly delays it. Now it'll run after interval + interval (first pass) + interval (actual cleaning). If interval is 1hr then you'll wait 3 hours until regular cleanup deletes something.
On the other hand, since you moved sleeping out of synchronized methods it's fine if scheduled runs first. Next time it'll just find out that some files are already deleted. See if it'll happen according to the current code.

AlekseiS replied:
Also, think if jitter time parameter needs to be different (smaller) for a cleanup at a startup.

omegaga commented:
Address @AlekseiS 's comments. @electrum Would you take a look?

AlekseiS commented:
@omegaga please rebase on the latest version and make sure CI tests are passing.

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -330,10 +361,43 @@ void deleteOldCompletedTransactions()
+                .map(ShardMetadata::getShardUuid)
+                .collect(toSet());
+
+        Set<UUID> deletions = new HashSet<>();
```
Use `Sets.difference()`

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -298,6 +320,15 @@ public void startLocalCleanup()
+    @Managed
+    public void startForceLocalCleanup()
+    {
+        Set<UUID> local = getLocalShards();
```
Do this scan inside the job

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -330,10 +361,43 @@ void deleteOldCompletedTransactions()
+    }
+
+    @VisibleForTesting
+    synchronized void cleanLocalShardsAtStartup(Set<UUID> local)
```
Let's rename this and the log message since it's used for JMX. How about "immediately" instead of "at startup".

electrum commented on presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestShardCleaner.java:
```diff
@@ -189,6 +189,55 @@ public void testDeleteOldShards()
+        shardDao.insertShardNode(shard1, node1);
+        shardDao.insertShardNode(shard3, node2);
+
+        // advance time within clean time
```
We don't need to advance time at all, since the clean code doesn't depend on time.

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -251,15 +252,13 @@ private void scheduleBackupCleanup()
+            waitJitterTime();
+            runLocalCleanupAtStartup(local);
+        });
         scheduler.scheduleWithFixedDelay(() -> {
```
Add a newline here, to separate these two independent actions

omegaga commented:
Address comments from @electrum . Travis breaks in flaky tests but look unrelated (one issue is as @jessesleeping mentioned in #7689 )

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -298,6 +321,15 @@ public void startLocalCleanup()
+    public void startLocalCleanupImmediately()
+    {
+        scheduler.submit(() -> {
+            Set<UUID> local = getLocalShards();
```
No need for a variable here

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -330,10 +362,38 @@ void deleteOldCompletedTransactions()
+                .map(ShardMetadata::getShardUuid)
+                .collect(toSet());
+
+        // mark all local files that are not assigned to deletions
```
"mark" is not accurate: we are just computing the list. We can say
```java
// only delete local files that are not assigned

electrum commented on presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java:
```diff
@@ -330,10 +362,38 @@ void deleteOldCompletedTransactions()
+
+        // mark all local files that are not assigned to deletions
+        Set<UUID> deletions = Sets.difference(local, assigned);
+        if (deletions.isEmpty()) {
```
Let's take this out so that so that we always print a log message. This is especially useful when run manually from JMX, so that the admin knows zero things were cleaned vs something went wrong.

omegaga commented:
Address comments from @electrum

electrum commented:
Merged, thanks!
